### PR TITLE
[wasm] Add flag WASM_THREAD_POOL_SIZE 

### PR DESCRIPTION
**Pull request by axinging**
This flag indicates how many threads can be used in multi-thread mode, it helps us understand the performance difference between wasm and webgpu/webgl.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/4942)

### CODE REVIEW

**Comment by axinging** [comment]
@qjia7 @xhcao @haoyunfeix @gyagp, 
Add @hujiajie.

File: tfjs-backend-wasm/src/backend_wasm.ts [hunk 1]
```diff
@@ -44,6 +44,12 @@ export class BackendWasm extends KernelBackend {
   constructor(public wasm: BackendWasmModule) {
     super();
     this.wasm.tfjs.init();
+    // Register the used thread pool size flag. Done it here to avoid circular
+    // import.
+    env().registerFlag('WASM_THREAD_POOL_SIZE', () => {
```
Comment (qjia7) [inline tfjs-backend-wasm/src/backend_wasm.ts:49]: Usually, we register flag in `flags_wasm.ts`. And call `env().set('WASM_THREAD_POOL_SIZE', this.getThreadPoolSize());`. But I am not sure if it is always the rules.

**Review by qjia7 — COMMENTED** [review]
@lina128 @jinjingforever take a look, thanks.

File: tfjs-backend-wasm/src/backend_wasm.ts [hunk 2]
```diff
@@ -347,8 +358,10 @@ export async function init(): Promise<{wasm: BackendWasmModule}> {
               'number',  // size
               'number',  // memoryOffset
             ]),
-        disposeData: module.cwrap('dispose_data', voidReturnType, ['number']),
-        dispose: module.cwrap('dispose', voidReturnType, []),
+        getThreadPoolSize:
+            module.cwrap('get_thread_pool_size', 'number', [voidType]),
```
[the remark this answers is not in the document]
Comment (axinging) [reply tfjs-backend-wasm/src/backend_wasm.ts]: Yes, this also works.

File: tfjs-backend-wasm/src/backend_wasm.ts (hunk 1 above)
Comment (axinging) [reply tfjs-backend-wasm/src/backend_wasm.ts:49]: Currently backend_wasm.ts  deppends on flags_wasm.ts.
If we put those code in flags_wasm.ts, flags_wasm.ts will also depends on  backend_wasm.ts, the test case will pass, but the yarn lint will fail due to  circular imports.

**Review by jinjingforever — APPROVED** [review]

File: tfjs-backend-wasm/src/cc/backend.cc [hunk 3]
```diff
@@ -90,6 +91,11 @@ EMSCRIPTEN_KEEPALIVE
 #endif
 void init() { xnn_initialize(nullptr); }
 
+#ifdef __EMSCRIPTEN__
+EMSCRIPTEN_KEEPALIVE
+#endif
+const size_t get_threads_count() { return backend::threads_count; }
```
Comment (huningxin) [inline tfjs-backend-wasm/src/cc/backend.cc:97]: FYI. There is a pthreadpool API to query the number of threads in a thread pool

https://github.com/Maratyszcza/pthreadpool/blob/master/include/pthreadpool.h#L86

File: tfjs-backend-wasm/src/backend_wasm.ts [hunk 4]
```diff
@@ -44,6 +44,12 @@ export class BackendWasm extends KernelBackend {
   constructor(public wasm: BackendWasmModule) {
     super();
     this.wasm.tfjs.init();
+    // Register the used thread pool size flag. Done it here to avoid circular
+    // import.
+    env().registerFlag('WASM_THREAD_POOL_SIZE', () => {
+      return this.getThreadPoolSize();
```
[the remark this answers is not in the document]
Comment (axinging) [reply tfjs-backend-wasm/src/backend_wasm.ts]: rename done

File: tfjs-backend-wasm/src/cc/backend.cc (hunk 3 above)
Comment (axinging) [reply tfjs-backend-wasm/src/cc/backend.cc]: done.

**Review by lina128 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/4942)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @axinging, @gyagp, @mattsoulanille, and @qjia7)

___
*[tfjs-backend-wasm/src/backend_wasm.ts, line 49 at r4](https://reviewable.io/reviews/tensorflow/tfjs/4942#-MYuhSdu4akHrWmWXMfY:-MYuhSdu4akHrWmWXMfZ:buabq57) ([raw file](https://github.com/tensorflow/tfjs/blob/042c9e63f2b20417e148aa6e2b182821c8114b2a/tfjs-backend-wasm/src/backend_wasm.ts#L49)):*
> ```XML
>     // Register the used threads count flag. Done it here to avoid circular
>     // import.
>     env().registerFlag('WASM_THREADS_COUNT', () => {
> ```

Is this flag tunable?

**Comment by axinging** [comment]
> > _[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/4942)_ status: :shipit: complete! 1 of 1 approvals obtained (waiting on @axinging, @gyagp, @mattsoulanille, and @qjia7)
> 
> _[tfjs-backend-wasm/src/backend_wasm.ts, line 49 at r4](https://reviewable.io/reviews/tensorflow/tfjs/4942#-MYuhSdu4akHrWmWXMfY:-MYuhSdu4akHrWmWXMfZ:buabq57) ([raw file](https://github.com/tensorflow/tfjs/blob/042c9e63f2b20417e148aa6e2b182821c8114b2a/tfjs-backend-wasm/src/backend_wasm.ts#L49)):_
> 
> > ```
> >     // Register the used threads count flag. Done it here to avoid circular
> >     // import.
> >     env().registerFlag('WASM_THREADS_COUNT', () => {
> > ```
> 
> Is this flag tunable?

Current this flag is untuable. The set pthread pool size logic is not done yet.

**Comment by axinging** [comment]
I think this change is ready now. Do you have any ideas?
As to the set thread pool size, I will work on it in another PR.

**Comment by axinging** [comment]
I will close this, and will provide a better solution for this.

**Pull request closed without merging**